# <br><br><span style="color:rebeccapurple">Python Fundamentals - Part 1</span>

## <br><br><span style="color:teal">0.0 Coding Notebooks
### This is a Jupyter Notebook.
Python coding notebooks are great for organizing your code alongside headers and text. They show the output of each code **snippet** directly below the code, which makes them great for debugging and creating visualizations.
<br><br>The alternatives to using a notebook are:
- **Interactive programming**: through a **shell** (interpreter, console) one line at a time
- **Batch programming**: running a whole **script** (a plain text file containing one to many lines of code)
- With the help of a GUI (graphical user interface). GUIs for coding are called **IDEs (Integrated Development Environments)**.
    - Some Python IDES: Spyder, PyCharm, Jupyter Lab, Google Colab, VS Code. Databricks and other compute environments also have their own IDEs built in.
    - IDEs provide helpful features like tab-complete when you're typing a variable name, automatically adding a closing quotation mark when you start typing a string, and color coding of different code components. You can usually customize or turn off these features as you like.

*The Jupyter Notebook we're working in now can be used for both interactive (one line at a time) and batch programming (many lines at a time).*

#### <br>What is Jupyter?
Jupyter is a non-profit, open source organization that created the original coding notebooks. Because they are **open source**, meaning they make their code available to all, almost all Python notebooks today are built with code that was written for Jupyter notebooks.

#### <br>More tips for using Jupyter notebooks
- Everyone's notebook will look a little different, depending on if you are running it locally through Jupyter Lab or another IDE on your computer, or on a cloud compute environment like Google Colab, GitHub Codespaces, or Databricks. You will also see small differences between operating systems (Mac or a PC) and browsers (Chrome, Safari, Firefox, Edge, etc.).
- Jupyter notebooks automatically save changes as you work.
- **AI coding assistants** are now turned on by default in most notebooks. The specific LLM will vary, depending on your IDE. For this workshop, you are welcome to try it out during breaks, but I think you are here to learn how to code, so you probably don't want to cheat on today's exercises. But your learning is in your hands! You can turn off the AI features for this workshop, and we will go over that now.

### How to use this notebook
Notebooks contain **code cells** and **text cells**.

Text cells like this one don't contain code. They are written in a language called **Markdown**.

You can change the cell type (and sometimes the programming language) from *Code* or *Python* to *Text* or *Markdown*.

#### To run a code cell:
Click in the code cell and either click on the "play" arrow, or type shift+enter (or shift+return on a Mac). Go ahead and run the cell below.

In [ ]:
print("Hello World!")

Hello World!


<br>When you hit shift+enter in a code cell, you **run** the code.
<br>If something gets printed out to the screen after you run the code, we say it is **returned**.

#### Working with markdown cells
If you accidentally double click in a markdown cell, it won't cause any problems. Go ahead and double click on this text. Then you can run the cell, just like you would with a code cell by hitting shift+enter, to return it to looking like nicely formatted text.

If you want to jot down a quick note to yourself as we work through these notebooks, you can add a new Markdown cell and type your text.

#### Adding a new cell
As we go, you may want to add a new code cell to try something out, or a new text cell to write down a note. In most notebooks, you should see a `+` sign to add a new cell.

### <br><br><span style="color:red">EXERCISE
Add a new code cell and a new text cell below this text cell.

## <br><br><span style="color:teal">0.1 Python
Python is an **open source** programming language. This means the source code is public, which means the language is free for anyone to use. It means anyone can develop their own Python package.
<br><br>Python is **maintained** (made available, checked for bugs) and improved (adding new features, changing fundamental operations) by a non-profit at Python.org called the Python Software Foundation (PSF). PSF is sponsored by all the major tech companies who rely on Python, which means those companies often pay a staff member some amount of time to help maintain Python. Volunteers are also a big support.
<br><br>Python runs behind the scenes in many tech tools (like your Macbook) and software programs.

### Python is a language.
(Not a software program.)

#### Python is:
**50% Syntax**
- Which words to use
- Punctuation
- Order
- Indentation (tabs and returns mean something in Python!)
- Shortcuts

**50% Logic**
- Which tools to use in which order
- Creative
- Specific to one problem

<br>Try not to worry if you don't understand everything today. You'll get more practice and time to think (and overthink!) as the workshop progresses. I will label some concepts we cover as **<span style="color:crimson">LOGIC**. LOGIC skills are very helpful for piecing together your own code, but if you are struggling on a LOGIC section, don't worry too much - you can focus on learning the syntax and come back to thinking about logic when you are writing your own code.

<br>*If you are coming from another coding language,* please remember that coding languages, like spoken languages, are not directly translated. You may think two things work the same way, but then find out they don't in a very inconvenient way! It is better to come to this workshop with an open mind.

## <br><br><span style="color:teal">1.0 Part 1: Objects and Functions

#### The two main concepts of Python.
- **Object**: a particular piece of data (like a noun)
- **Function**: something you can do to/with an object (like a verb)

<br>You can use certain functions with certain classes of objects.
<br>
<br>Some functions are shared between classes of objects, and some are not.

<br><br><br>Let's consider this example. I have a dog named Molly, and I have a bicycle. There are some things I can do to both Molly and my bicycle, and some things I can't do to both:

## 🐕 Molly

- Feed Molly
- Pet Molly
- Draw Molly

## 🚲 My bicycle
- Pedal my bicycle
- Pet my bicycle (limited use)
- Draw my bicycle

<br><br><br>The things I can do to Molly are actually things I can do to any dog, and the things I can do to my bicycle are things I can do to any bicycle:

## 🐕 Molly (class dog)

- Feed a dog
- Pet a dog
- Draw a dog

## 🚲 My bicycle (class bicycle)
- Pedal a bicycle
- Pet a bicycle (limited use)
- Draw a bicycle

<br><br>Python example: We can do different things to the number three, depending on which type of object it is.

## 🔢 &nbsp; 3 (class integer)

- Divide (3 / 3 = 1)
- Convert 3 to 3.0
- Add (3 + 3 = 6)

## 📝 &nbsp; "3" (class string)
- Replace ("3" becomes "1")
- Convert "3" to 3.0
- Add ("3" + "3" = "33")

### <br><br><br>Objects in Part 1
- Integer (whole number) 3
- Float (number with digits following a decimal point) 3.0
- String (text or characters) “3” or “three”
- Boolean (True or False)
- *Assigning objects to variables*
- *Indexing to get only part of an object*

### Functions in Part 1
- Basic operators: + - * /
- Functions to convert between data types
- String functions
- *Coding story problems*

<br><br><br><span style="color:magenta">**LET'S CODE!**

## <br><br><span style="color:teal">1.1 Arithmetic operators with integers and floats

In [ ]:
4 + 2

6

In [ ]:
6 - 1

5

In [ ]:
100 * 35

3500

In [ ]:
40 / 6

6.666666666666667

To calculate an exponential:

In [ ]:
4 ** 2

16

<br><br>You might notice that I've included spaces around the operators:
<br>`4 + 2`
<br>instead of
<br>`4+2`
<br><br>This is best practice in Python because it is more readable. The math would work without spaces, but it's a good idea to get into the habit of writing "Pythonic" code now!

#### <br><br>More division operators:

In [ ]:
40 / 5

8.0

We gave it **integers** (whole number objects) and it returned a **float**. The `/` division operator always returns a float.

<br>Get just the integer, without the remainder (floor division):

In [ ]:
40 // 6

6

<br>Get just the remainder (modulo):

In [ ]:
40 % 6

4

<br>Like many things you'll learn this week, I'm not expecting you to memorize the operator for returning a remainder. If this is something you use a lot already in your research, you might memorize it now. But if not, I just want you to hear about it so that if you ever need it months from now, you might remember that there is a way to do it so that you can google it and find the right code.</span>

<br>More practice:

In [ ]:
40 / 5

8.0

In [ ]:
40 // 5

8

In [ ]:
40 % 5

0

#### <br><br>Python follows the PEMDAS order of operations: (https://thehelloworldprogram.com/python/python-operators-order-precedence/)

In [ ]:
40 / 5 + 2

10.0

In [ ]:
40 + 2 * 6 - 1

51

In [ ]:
(40 + 2) * (6 - 1)

210

Why did some lines of code return integers (51, 210) and one returned a float (10.0)?

### <br><br><span style="color:red">Exercise: Arithmetic Operators

Try out some math of your own:

## <br><br><span style="color:teal">1.2 Comparison operators and booleans

These return a **boolean** object: `True` or `False`.

In [ ]:
3 > 2

True

In [ ]:
3 < 2

False

In [ ]:
3 >= 2

True

In [ ]:
3 <= 2

False

To check if two things are **equal**, Python uses a double equals sign:

In [ ]:
3 == 3

True

To check if two things are **not equal**, Python uses `!=`:

In [ ]:
3 != 3

False

In [ ]:
3 != 2

True

### <br><br><span style="color:red">Exercise: Comparison Operators

Write code to find out if 1.00486 is greater than 1.000918.

In [ ]:
1.00486 > 1.000918

True

Write code to find out if 6 is equal to 6.0.

In [ ]:
6 == 6.0

True

## <br><br><span style="color:teal">1.3 Logical operators

These also return a **boolean**.

`and` means both sides must be True to return True:

In [ ]:
3 < 4 and 4 == 3

False

`or` means only one side must be True to return True:

In [ ]:
3 < 4 or 4 == 3

True

`not` is another logical operator. It reverses the boolean that follows it:

In [ ]:
not 2 == 2

False

Python uses the lowercase words for `and`, `or`, and `not`. Newer versions of Python may be able to interpret `&`, `|`, and `!`, but not all packages you use will work with those symbols, so it's best to get used to using the lowercase words.

### <br><br><span style="color:red">Exercise: Logical Operators

Before running these cells, try to guess if they will return `True` or `False`:

In [ ]:
not 2 > 3 and not 2 == 2

False

In [ ]:
not 2 > 3 or not 2 == 2

True

In [ ]:
3 == 3 and not 3 != 3

True

## <br><br><span style="color:teal">1.4 String objects

A **string** is an object kind of like text.
<br><br>It is contained in double `"` or single `'` quotes (you can choose which you want to use, but I use double quotes, so that is what we will use for this workshop). Again - both work just fine!

In [ ]:
"Hello World"

'Hello World'

<br><br>Let's use our first **function**! The `print()` function can be used on almost all classes of objects:

In [ ]:
print("Hello World")

Hello World


In [ ]:
print(334)

334


<br>Strings can contain digits:

In [ ]:
print("Hello World 2")

Hello World 2


### <br><br>Special Characters

Strings can contain **special characters**. These start with a backslash (on my keyboard, the backslash key is found above my return key). The backslash can also be called the **escape character**. <br><br>The most commonly used special characters are new lines `\n`:

In [ ]:
print("Hello World\nHow are you?")

Hello World
How are you?


...and tabs `\t`:

In [ ]:
print("Hello World\tHow are you?")

Hello World	How are you?


<br><br>If you simply **run** the string, it will return the string how it looks to the computer:

In [ ]:
"Hello World\nHow are you?"

'Hello World\nHow are you?'

...but if you use the `print()` function, it **interprets** the special characters and returns the string like this:

In [ ]:
print("Hello World\nHow are you?")

Hello World
How are you?


<br><br>If you ever need to include an actual backslash in your string and you don't want it to be interpreted as a special character, you can use a double backslash:

In [ ]:
print("always\never")

always
ever


In [ ]:
print("always\\never")

always\never


### <br><br>Using quotation marks in strings
You can include single and double quotation marks in your strings:

In [ ]:
print("Here's an example.")

Here's an example.


In [ ]:
print('Hello "World"')

Hello "World"


In the first example, we used a single quotation mark (apostrophe) inside double quotation marks. In the second example, we used double quotation marks inside single quotation marks.

<br>What if you want to include both single and double quotation marks in a string? You can use the escape character (the backslash) to let the computer know to interpret the following character as part of the string:

In [ ]:
print("Here's the \"example\".")

Here's the "example".


### <br><br><span style="color:red">Exercise: Strings and Special Characters

Write code to print out the following message (it should look exactly as it is shown here):
<br><br>This is line 1.
<br>This is line 2.

In [ ]:
print("This is line 1.\nThis is line 2.")

This is line 1.
This is line 2.


### <br><br>Combining strings

Some **arithmetic operators** work with strings. To combine two strings, we use the `+` sign:

In [ ]:
print("Hello" + "World")

HelloWorld


In [ ]:
print("Hello " + "World")

Hello World


## <br><br><span style="color:teal">1.5 Python errors

What happens if you forget the quotes around a string?

In [ ]:
print(Hello World)

SyntaxError: invalid syntax. Perhaps you forgot a comma? (4293340409.py, line 1)

<br><br>Our first error!! This is a common error - the `SyntaxError`. It often means that we forgot to type something.

<br><br>What happens if you spell a function wrong?

In [ ]:
pint("Hello World!")

NameError: name 'pint' is not defined

<br><br>A `NameError` is another common error - it often means we've mispelled a function or variable.

<br><br>What error do you think we will get if we run this line of code?

In [ ]:
print("Hello World)

SyntaxError: unterminated string literal (detected at line 1) (2771583741.py, line 1)

## <br><br><span style="color:teal">1.6 Object classes interacting

<br>The `type()` function tells us the object type of an object:

In [ ]:
type("Hello World")

str

In [ ]:
type(9)

int

In [ ]:
type(9.9)

float

In [ ]:
type("9.9")

str

### <br>Converting between object classes
When it makes sense, we can convert objects from one type to another, using the appropriate function:

In [ ]:
float(9)

9.0

In [ ]:
float("5")

5.0

In [ ]:
int("5")

5

In [ ]:
int(5.7)

5

*Notice it doesn't round when it converts a float to an integer, it just drops off everything after the decimal. You can use the `round()` function to round, but we'll cover that a bit later.*

In [ ]:
str(5.005)

'5.005'

In [ ]:
float("Hello World")

ValueError: could not convert string to float: 'Hello World'

<br>Our third type of error! This error usually means you passed an argument to a function that it doesn't recognize. But don't spend any brainpower trying to memorize this error. Luckily the error message gives us information about what we did wrong.

### <br><br>Object classes interacting

Before running each of these cells, try to guess what they will return:

In [ ]:
4 * 4.0

16.0

In [ ]:
int(4 * 4.0)

16

In [ ]:
"Hello World" * 4

'Hello WorldHello WorldHello WorldHello World'

In [ ]:
5 + True

6

In [ ]:
5 + False

5

*Note: `True` evaluates to 1, `False` evaluates to 0*

### <br><br><span style="color:red">Exercise: Object Classes Interacting

Run the next line of code:

In [ ]:
print("Hello World" + 4)

TypeError: can only concatenate str (not "int") to str

<br>A `TypeError` means you tried to do something that doesn't work with the type of object you gave it. Sometimes `ValueError`s and `TypeError`s get thrown for similar errors.

<br>Use a *function* to modify the code below to create a string that says "Hello World4".

In [ ]:
print("Hello World" + str(4))

Hello World4


## <br><br><span style="color:teal">1.7 Variables

Variables allow you to store an object to use it later.

In [ ]:
x = 3

Nothing is returned when you assign a variable. You can see the value of a variable by running the variable name or using the `print()` function

In [ ]:
x

3

In [ ]:
print(x)

3


<br>The single `=` is called an **assignment operator**.
<br>Everything on the right side of the assignment operator is calculated first, and then that value is stored as the variable on the left of the assignment operator.

In [ ]:
x = 3 - 1

In [ ]:
x

2

In [ ]:
y = x

In [ ]:
y

2

In [ ]:
x = x ** y

In [ ]:
x

4

In [ ]:
y

2

In [ ]:
rabbit = 50

Variables can be used just like the objects they represent:

In [ ]:
rabbit - 20

30

In [ ]:
rabbit

50

In [ ]:
greeting = "Hello you!"

In [ ]:
greeting

'Hello you!'

In [ ]:
print(greeting)

Hello you!


### <br><br><span style="color:red">Exercise: Creating Variables

Create a variable called `favorite_color` and assign it to a string that contains your favorite color. Then write code to print the string you saved as `favorite_color`.

In [ ]:
favorite_color = "blue"

In [ ]:
print(favorite_color)

blue


### <br><br>Variable names

#### Python Rules
- No spaces
- Case matters: x is not X
- Must start with a letter
- Should be meaningful (don't save your number as `rabbit` unless it represents the number of rabbits)
- For longer variable names: separate_with_underscores

## <br><br><span style="color:teal">1.8 Functions

We've been learning a lot about objects. Let's learn more about functions!

<br>**Parentheses immediately follow a function name (no space).**

In [ ]:
print("Hello World")

Hello World


The information inside the parentheses is called an **argument**.
<br>We say we **passed** the argument "Hello World" to the print function.

<br><br>Some functions take multiple arguments. The arguments must be separated by a comma:

In [ ]:
round(4.5874, 2)

4.59

We'll learn more about the round() function later today. It takes two arguments: an integer or a float to be rounded, and the number of places after the decimal point that the first number should be rounded to.

### <br><br>Two types of functions in Python

1. A true **function**.
- Stands on its own.
- Takes an object as an **argument**.
- These types of functions usually do something **with** the object.
- Can usually be used on multiple object classes.

In [ ]:
my_name = "Colby"

In [ ]:
print(my_name)

Colby


In [ ]:
print(1045)

1045


In [ ]:
len(my_name)

5

<br><br>2. **Method**
- Always follows an object.
- Some take arguments and some don't.
- These types of functions usually do something **to** the object.
- Methods are specific to one object class.

In [ ]:
my_name = "Colby"

In [ ]:
my_name.upper()

'COLBY'

<br>Here we can see that methods are specific to one object class. A **string** can be turned into uppercase, but an **integer** cannot:

In [ ]:
1045.upper()

SyntaxError: invalid decimal literal (4168500355.py, line 1)

<br>Here is an example of a method function that takes multiple arguments, separated by a comma:

In [ ]:
my_name.replace("C", "c")

'colby'

We will learn more about the `.replace()` method after the next exercise.

### <br><br><span style="color:red">Exercise: Functions

Create a new variable called `full_name`. It should store the string of your own full name.

In [ ]:
full_name = "Colby Witherup Wood"

In [ ]:
print(full_name)

Colby Witherup Wood


Write code to find out the length of your full name.

In [ ]:
len(full_name)

19

Did the length function count spaces when it calculated the length of your name?

*When Python developers create Python functions, they have to make choices about how the functions will work. For example, should the length function count spaces or not? It's important to test out functions to see how they work if you are unsure or if you are getting unexpected results.*

## <br><br><span style="color:teal">1.9 String functions
Let's learn more about how strings work.

In [ ]:
greeting = "Hello you!"

In [ ]:
print(greeting)

Hello you!


In [ ]:
len(greeting)

10

<br>The `replace()` function will replace one character with another.

In [ ]:
greeting.replace("!", ".")

'Hello you.'

In [ ]:
greeting.replace(" ", "")

'Helloyou!'

`""` is referred to as an **empty string**.
<br><br>

In [ ]:
print(greeting)

Hello you!


<br>Even though we ran code to change our greeting, the greeting variable still **points** to the original greeting. `greeting` did not change when we used the method functions even though a changed version of `greeting` was **returned**.

<br>To change the greeting, we would have to **reassign** the variable name:

In [ ]:
greeting = greeting.replace("!", ".")

In [ ]:
print(greeting)

Hello you.


In [ ]:
big_greeting = greeting.upper()

In [ ]:
print(big_greeting)

HELLO YOU.


In [ ]:
small_greeting = greeting.lower()

In [ ]:
print(small_greeting)

hello you.


In [ ]:
my_name = "Colby"

In [ ]:
greetMe = greeting.replace("you", my_name)
print(greetMe)

Hello Colby.


*Note: You might have noticed that I included two lines of code in that code cell. Jupyter Noteboooks can run multiple lines of code in a single code cell.*

## <br><br><span style="color:teal">1.10 Indexing strings

We can **index** strings to return a **substring**. A substring can be one or more characters long.

#### <br><span style="color:magenta">**In Python indexing, we start counting with 0.**

#### H &nbsp; &nbsp; E &nbsp; &nbsp; L &nbsp; &nbsp; L &nbsp; &nbsp; O &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; W &nbsp; &nbsp; O &nbsp; &nbsp; R &nbsp; &nbsp; L &nbsp; &nbsp; D &nbsp; &nbsp; !
#### 0 &nbsp; &nbsp; 1 &nbsp; &nbsp; 2 &nbsp; &nbsp; 3 &nbsp; &nbsp; 4 &nbsp; &nbsp; 5 &nbsp; &nbsp; 6 &nbsp; &nbsp; 7 &nbsp; &nbsp; 8 &nbsp; &nbsp; 9 &nbsp; &nbsp; 10 &nbsp; &nbsp; 11

<br>To return a single character, we type our string followed immediately by square brackets (no space). Inside the brackets, we pass the index position for the character.

In [ ]:
"Hello World!"[0]

'H'

In [ ]:
"Hello World!"[4]

'o'

#### <br><br>H &nbsp; &nbsp; E &nbsp; &nbsp; L &nbsp; &nbsp; L &nbsp; &nbsp; O &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; W &nbsp; &nbsp; O &nbsp; &nbsp; R &nbsp; &nbsp; L &nbsp; &nbsp; D &nbsp; &nbsp; !
#### 0 &nbsp; &nbsp; 1 &nbsp; &nbsp; 2 &nbsp; &nbsp; 3 &nbsp; &nbsp; 4 &nbsp; &nbsp; 5 &nbsp; &nbsp; 6 &nbsp; &nbsp; 7 &nbsp; &nbsp; 8 &nbsp; &nbsp; 9 &nbsp; &nbsp; 10 &nbsp; &nbsp; 11

To take a longer **substring**, you start with the position of the first character you want included, then a colon, and end with the position *1 past* the last character you want.

To index "Hello":

In [ ]:
"Hello World!"[0:5]

'Hello'

<br><br>If this seems strange, it might help you to know what is happening behind the scenes. Python is actually labeling the invisible spot *between* each character in our string:

#### &nbsp; &nbsp; H &nbsp; &nbsp; E &nbsp; &nbsp; L &nbsp; &nbsp; L &nbsp; &nbsp; O &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; W &nbsp; &nbsp; O &nbsp; &nbsp; R &nbsp; &nbsp; L &nbsp; &nbsp; D &nbsp; &nbsp; !
#### 0 &nbsp; &nbsp; 1 &nbsp; &nbsp; 2 &nbsp; &nbsp; 3 &nbsp; &nbsp; 4 &nbsp; &nbsp; 5 &nbsp; &nbsp; 6 &nbsp; &nbsp; 7 &nbsp; &nbsp; 8 &nbsp; &nbsp; 9 &nbsp; &nbsp; 10 &nbsp; &nbsp; 11 &nbsp; 12

So now it's easier to see why "Hello" would be indexed with [0:5].<br><br><br><br>

In [ ]:
greeting = "Hello you!"

In [ ]:
greeting[0:5]

'Hello'

In [ ]:
hello = greeting[0:5]
print(hello)

Hello


In [ ]:
hello[2]

'l'

*Remember, in Python we start indexing at 0, so the third letter in our string is indexed as 2.*

### <br><br>Negative indexing
We can also start indexing from the right end of the string. The character in the farthest right position can be indexed as -1. We call this **negative indexing**.

In [ ]:
hello[-1]

'o'

In [ ]:
hello[-2]

'l'

In [ ]:
hello[0:2]

'He'

*Remember, our variable is `hello`, but the string it has stored is "Hello" with a capital "H".*

### <br><br>Indexing shortcut for substrings
If we don't give any number after the `:`, Python will continue to the end of the string:

In [ ]:
hello[3:]

'lo'

If we don't give any number before the `:`, Python will start at the beginning of the string:

In [ ]:
hello[:2]

'He'

<br>We can also use negative numbers to index substrings:

In [ ]:
hello[2:-1]

'll'

In [ ]:
hello[-3:-1]

'll'

In [ ]:
"Hello"[-3:]

'llo'

### <br><br><span style="color:red">Exercise: Indexing strings

Run the line of code below to store the variable `movie`.

In [ ]:
movie = "TENET"

Write code to index the first character in the variable `movie`.

In [ ]:
movie[0]

'T'

Use negative indexing to return the last character in the variable `movie`.

In [ ]:
movie[-1]

'T'


<br>Is the first letter of `movie` the same as the last letter of `movie`? Combine the indexing statements you just wrote with a comparison operator to see if the two letters are equal.

In [ ]:
movie[0] == movie[-1]

True

*Your code should return a boolean: True or False.*

<br>Run the cell below to store the plot summary of the movie "Tenet".

In [ ]:
movie_plot = "The film follows a former CIA agent who is recruited \
              into a secret organization, tasked with tracing the origin \
              of objects that are traveling backward through time and \
              their connection to an attack from the future to the present."

*To break up a long string of text into shorter lines, you can put a `\` (backslash) at the end of each line.*

<br>Write code to index the substring "objects that are traveling backward through time" from the `movie_plot` variable.

In [ ]:
movie_plot[143:-80]

'objects that are traveling backward through time'

## <br><br><span style="color:teal">1.11 Modules (aka packages or libraries)

By default, only a handful of the most commonly used functions and object classes are loaded when you start Python.
<br><br>This saves memory so that your Python scripts can run as quickly as possible.
<br><br>The object classes and functions that are loaded automatically with Python are called **built-in**.

#### <br>Modules
- Modules are groups of related functions and object classes
- They are not automatically loaded when you start Python
- You must **import** the module every time you start Python (or in every Jupyter Notebook)
- If you do not have the module on your computer or in your cloud environment, you must first **install** it (only once)
- If you are working in an IDE, many of the most popular Python packages are ready to **import** (they are pre-installed). To use other packages, you must **install** the package first, and then **import**.

<br>Today we'll learn how to **import** packages and work with a couple functions from a package that is included in every installation of Python but isn't automatically loaded when you start Python.
<br><br>Later in the workshop we'll talk briefly about **installing** packages.

We are going to use some functions from the module `math`.

You should usually import any modules at the top of your Jupyter notebook, even if you don't use them until much later in the notebook. This is so that people can see which modules they need as soon as they open the notebook, in case they need to **install** the module on their computer. *We're breaking this rule today and loading the module here.*

Run this cell to import the `math` module:

In [ ]:
import math

<br><br>To use a function from an imported module, you type the name of the module followed by `.` followed by the name of the function. Let's use `math.sqrt()` to find the square root of a number:

In [ ]:
math.sqrt(16)

4.0

In [ ]:
math.sqrt(15.5)

3.9370039370059056

<br>The `math` module contains many mathematical functions. It also contains a few objects, like `math.pi`, which is the value of pi, so that you don't have to memorize it. You know it's an object and not a function because it doesn't have parentheses at the end.

In [ ]:
radius = 6
area_of_a_circle = math.pi * radius**2
print(area_of_a_circle)

113.09733552923255


*Remember that `**` is used to find an exponent like a square.*

### <br><br><span style="color:red">Exercise: Use an object from a module

The equation for calculating the circumference of a circle is $2 \times \pi \times radius$. Write code to calculate the circumference of a circle with a radius of 10.

In [ ]:
radius = 10
circumference = 2 * math.pi * radius
print(circumference)

62.83185307179586


### <br><br>Package repositories and package managers
Package **repositories** are collections of Python packages that are hosted online for you to use.
<br><br>To download a package from a repository, you have to use a **package manager**, which is a software tool that does three things:
- installs a package from the repository
- connects the package to your version of Python behind the scenes
- handles any **dependencies** (many packages are built on top of other packages, so the package manager will make sure you install those packages too)

<br>There are a few big package repositories. **The big repositories all host packages that have already been checked for any malware.** This is important because packages are computer code - they are designed to run code on your computer - so you should only use a trusted source. Most packages on the big repositories have also been checked for bugs, but some repositories specifically host newer packages that are still in development.
<br><br>There are currently three big package managers that we will discuss.

#### <br>PyPI (Python Package Index) and pip (Python Installs Packages)
**PyPI** is a Python package repository, an online collection of free Python modules that are written and vetted by Python users. PyPI uses the **pip** package manager.
<br><br>PyPI is maintained by Python.org. All of the most popular Python packages are available from the PyPI package repository. All packages available on PyPI should already be vetted and checked for bugs and malware.

#### <br>conda, conda-forge, and bioconda
conda is the name of both a Python package **repository** and the name of the associated **package manager**. All packages on conda will have been vetted and checked for bugs and malware. conda-forge is another package repository for less developed packages that are still in development (but will have been checked for malware). Bioconda is yet another package repository for Biomedical packages. conda, conda-forge, and Bioconda are all maintained by Anaconda and all use the conda package manager for installations and updates.
<br><br>**Anaconda is a for-profit company. The conda package repository is free only for educational uses and not-for-profit research. If your research is for commercial profit, or has any potential of profit, do not use conda.** conda-forge is free for everyone. mamba is a great alternative to conda.

#### <br>mamba
**mamba** is a package manager and an alternative to conda. It pulls packages from the conda-forge repository among others. Like conda, it is often better at solving dependencies than pip.

<br>conda and mamba also allow you to create **environments**. Environments allow you to set up different versions of Python and associated packages in different environments that you can switch between. This can be very helpful if you are working with an old piece of software that needs an older version of Python, or if you have different requirements for different projects.

#### <br>Other packages
Other packages must be downloaded from the creator (such as from their GitHub site). This happens sometimes with packages developed by academics that you might find in a journal article. Note that these packages have not been checked for malware or bugs by a trusted package repository, or they could have incorrect calculations. They should be used with care.

## <br><br><span style="color:teal">1.12 Story problems: turning real world problems into Python code.

*When running this code, don't forget to run every code cell. If you get a `NameError` it might be because you didn't run a cell where a variable was defined.*

<br>I told the class I ate 12 bananas for lunch:

In [ ]:
bananas = 12

You saw me secretly eat 3 more bananas:

In [ ]:
bananas = bananas + 3

How many bananas did I eat?

In [ ]:
print(bananas)

15


Let's add some strings together to print the answer out as a complete sentence:

In [ ]:
answer = "Colby ate " + bananas + " bananas."

TypeError: can only concatenate str (not "int") to str

<br>Error! Let's check the type of the variable `bananas`.

In [ ]:
type(bananas)

int

We can change `bananas` to a string and save it as a new variable so that we can include it in our answer...

In [ ]:
bananas_str = str(bananas)
answer = "Colby ate " + bananas_str + " bananas."
print(answer)

Colby ate 15 bananas.


<br>...or we can change it to a string *inside* the code to create our answer:

In [ ]:
answer = "Colby ate " + str(bananas) + " bananas."
print(answer)

Colby ate 15 bananas.


<br><br>Our final code for this story problem looked like this:

In [ ]:
bananas = 12
bananas = bananas + 3
answer = "Colby ate " + str(bananas) + " bananas."
print(answer)

Colby ate 15 bananas.


<br>I also could have written the code like this, creating a new variable in the second line instead of reusing the variable name `bananas`:

In [ ]:
bananas = 12
real_bananas = bananas + 3
answer = "Colby ate " + str(real_bananas) + " bananas."
print(answer)

Colby ate 15 bananas.


<br>Or I could have written the code like this:

In [ ]:
bananas = 12
secret_bananas = 3
real_bananas = bananas + secret_bananas
answer = "Colby ate " + str(real_bananas) + " bananas."
print(answer)

Colby ate 15 bananas.


<br>Or I could have written the code like this:

In [ ]:
bananas = 12
print("Colby ate " + str(bananas + 3) + " bananas.")

Colby ate 15 bananas.


<br>These alternative ways of getting to the same answer are all acceptable. The difference between them involves when I decided to assign values to variables. There isn't always a right or wrong answer, but there are some guidelines to make your code more "Pythonic".

<br>In general, you want your code to be succinct but readable. You don't want to store too many variables in your computer's memory at one time, especially if they contain large objects like whole files (we'll talk about files and memory on Day 3). If you will never use the object again in the code, you might not want to bother assigning that object to a variable. Then again, if it makes it more readable to assign it to a variable, you should.

<br>For now, just focus on writing code that works! All of the options above would be acceptable.

### <br><br><span style="color:red">Exercise: Story Problems 1

My kitchen is 10 feet long by 10 feet wide. How many square feet is my kitchen? (The equation for area is length multiplied by width.) Finish the code below by assigning the area equation to the variable `area`:

In [ ]:
length = 10
width = 10
area = length * width
print("The kitchen is " + str(area) + " square feet.")

The kitchen is 100 square feet.


### <br><br><span style="color:red">Exercise: Story Problems 2

I scored 80 points on my first test and 88.5 points on my second test. What is my average test grade? (The equation for average test grade is the sum of all test scores divided by the number of tests.) Fill in the appropriate code below:

In [ ]:
test1 = 80
test2 = 88.5
average = (test1 + test2) / 2
print("Colby's average test grade is " + str(average) + ".")

Colby's average test grade is 84.25.


### <br><br><span style="color:red">Exercise: Story Problems 3

Last week I ran 10 miles. This week I ran the same as last week, plus 6 extra miles. How many miles did I run in total over both weeks? Write the code below to calculate the total number of miles. As a bonus, write the code to print out a complete sentence that says how many miles I ran.

In [ ]:
last_week = 10
this_week = last_week + 6
total = last_week + this_week
print("Last week I ran " + str(total) + " miles.")

Last week I ran 26 miles.


## <br><br><span style="color:teal">1.13 More about functions: default values and keyword arguments

<br>To review, some functions take only one argument.
<br>The `abs()` function returns the absolute value of a number:

In [ ]:
abs(-19.6)

19.6

<br>The `round()` function will round a float to a specified number of decimal places. It takes two arguments: the number to round and the number of decimal places.

In [ ]:
my_number = 631.890382720

In [ ]:
round(my_number, 4)

631.8904

In [ ]:
round(my_number, 2)

631.89

#### <br><br>Default values
Some function arguments have **default values**, so if you don't include the argument, it will use the default value. Let's try `round()` without specifying a value for the second argument:

In [ ]:
round(my_number)

632

So the default value is to round to the closest integer.

`round()` can also round to places to the left of the decimal point:

In [ ]:
round(18192948, -3)

18193000

### <br><br><span style="color:red">Exercise: Functions with default values

I am ordering 3 tacos for lunch. Each taco is $3.55. I would like to round my total up to the nearest dollar to leave a small tip. How much should I pay? Write the code below:

In [ ]:
taco = 3.55
lunch_bill = round(taco * 3)

In [ ]:
print(lunch_bill)

11


### <br><br><br>Keyword arguments
Some functions have more than one argument that has a default value. To change the default value, you have to pass the function a **keyword argument**. Let's see how this works.

The function `isclose()` is in the `math` module. Let's import the math module again, in case you closed and reopened this notebook during lunch.

In [ ]:
import math

<br><br>`math.isclose()` will tell you if two numbers are close. It returns a boolean.

In [ ]:
math.isclose(4, 5)

False

4 and 5 are not considered close.

In [ ]:
math.isclose(0.4, 0.5)

False

Still not close.

In [ ]:
math.isclose(0.0004, 0.0005)

False

Still not close.

In [ ]:
math.isclose(.4999999999, .5)

True

<br>*Relative tolerance* is the maximum allowed difference between a and b. The **default** relative tolerance is 1e-09.
<br><br>We can change the relative tolerance by passing the **keyword argument** `rel_tol`.

In [ ]:
math.isclose(4.03, 4.02, rel_tol=.01)

True

## <br><br><span style="color:teal">1.14 More about functions: Nested functions

You can also put functions inside other functions. The innermost function will be evaluated first, and the outermost function will be evaluated last.

In [ ]:
round(abs(-12.908))

13

In [ ]:
abs(round(-12.908))

13

In [ ]:
round(30 + 698, -1)

730

In [ ]:
round(30 + abs(-698), -1)

730

<br>The order of the nested functions matters. Look at the next two code cells and guess what will happen in each before you run the cell.

In [ ]:
abs(float("-6.342"))

6.342

In [ ]:
float(abs("-6.342"))

TypeError: bad operand type for abs(): 'str'

## <br><br><span style="color:teal">1.15 More about functions: Finding information
We can get information about a function by looking it up in the documentation <https://docs.python.org/3/library/index.html> or by coding the function name without the `()` followed by `?`:

In [ ]:
len?

Signature: len(obj, /)
Docstring: Return the number of items in a container.
Type:      builtin_function_or_method

### <br><br><span style="color:red">Exercise: More functions that work with integers and floats
The `math` package contains lots of functions that work with integers and/or floats.
<br><br>First, import the `math` package:

In [ ]:
import math

<br>Go to the documentation page for the `math` package: https://docs.python.org/3/library/math.html. Choose a function to test out. Does it work the way you thought it would? Are there any arguments that you can pass to the function?

## <br><br><span style="color:teal">1.16 More string functions

<br>We already learned these string functions:

In [ ]:
song = "We are the world"

In [ ]:
song.lower()

'we are the world'

In [ ]:
song.upper()

'WE ARE THE WORLD'

In [ ]:
song.replace("world", "weird")

'We are the weird'

In [ ]:
new_word = "wind"
song.replace("world", new_word)

'We are the wind'

<br>We also learned that these functions don't change the song:

In [ ]:
print(song)

We are the world


...and that we can save a changed version of the object by assigning a new variable, or reassigning the old variable:

In [ ]:
song = song.upper()
print(song)

WE ARE THE WORLD


### <br><br>More capitalization functions

We learned `upper()` and `lower()`. Run the code below to see what the `title()` and `capitalize()` methods do.

In [ ]:
song = "All along the watchtower"

In [ ]:
song.title()

'All Along The Watchtower'

In [ ]:
song = song.lower()
print(song)

all along the watchtower


In [ ]:
song.capitalize()

'All along the watchtower'

## <br><br><span style="color:teal">1.17 Removing characters from a string

In [ ]:
song = "All along the watchtower"

<br>The `lstrip()` method will remove a character or set of characters if it is the first character on the left side of the string:

In [ ]:
song.lstrip("a")

'All along the watchtower'

In [ ]:
song.lstrip("all")

'All along the watchtower'

In [ ]:
song.lstrip("w")

'All along the watchtower'

"w" is not the first character on the left side of the string, so nothing was removed.

<br>Note that the `.lstrip()` method is not for removing exact patterns. It will take any of the characters you pass as an argument in any order. Let's apply the same method and argument to all these strings:

In [ ]:
"all along the watchtower".lstrip("all")

' along the watchtower'

In [ ]:
"llama llama, red pajama".lstrip("all")

'ma llama, red pajama'

In [ ]:
"lalalalalatida".lstrip("all")

'tida'

<br><br>The `rstrip()` method will remove a character or set of characters if it is the first character on the right side of the string:

In [ ]:
song.rstrip("er")

'All along the watchtow'

<br>I use `rstrip()` all the time to remove new line characters from the end of lines in a file, or to remove trailing whitespace:

In [ ]:
"line in a file\n\n".rstrip("\n")

'line in a file'

In [ ]:
"line in a file      ".rstrip(" ")

'line in a file'

*Notice that it removed all the spaces on the right end of the string, even though we only passed it one space as the argument.*

#### <br><br>Removing a character everywhere it appears in a string
Use the `replace()` method we already learned and replace the character with an **empty string**:

In [ ]:
song.replace(" ", "")

'Allalongthewatchtower'

### <br><br><span style="color:red">Exercise: Removing characters from a string

Run the next line of code to save the variable `book`.

In [ ]:
book = "The Three Musk-eteers"

Write code to remove the "-" in the middle of `book`.

In [ ]:
book.replace("-", "")

'The Three Musketeers'

**Bonus exercise:** Write code to remove the "-" in the middle of the word "Musketeers" in `book2`, but not the "-" that divides the title from the subtitle.

In [ ]:
book2 = "The Three Musk-eteers - Part Two"

In [ ]:
book2.replace("k-e", "ke")

'The Three Musketeers - Part Two'

OR

In [ ]:
book2.replace("-", "", 1)

'The Three Musketeers - Part Two'

## <br><br><span style="color:teal">1.18 Stringing together method functions

You can string multiple method functions on the end of an object. The functions will be executed from left to right.

In [ ]:
song = "Let It Go"

In [ ]:
different_song = song.replace("Go", "Be").upper()
print(different_song)

LET IT BE


<br>The order of functions can change what is returned. Why do you think reversing the order of functions turns out the way it does?

In [ ]:
different_song = song.upper().replace("Go", "Be")
print(different_song)

LET IT GO


<br><br><br>Once you've made the song uppercase, "Go" does not exist in the string, so it won't be replaced. "Go" is already "GO".

<br>You can also string a method function before or after you index a substring. Again, the code will execute from left to right.

In [ ]:
song = "Let It Go"

In [ ]:
song[-2:].upper() + "!"

'GO!'

In [ ]:
song.upper()[-2:] + "!"

'GO!'

<br>*Side note, in case this ever happens to you, if you forget to put the parantheses on the end of your method function, you might not get an error, but you will see something like this:*

In [ ]:
song.lower

<function str.lower()>

## <br><br><span style="color:teal">1.19 String indexing and function practice

Earlier in the workshop you learned how to index a string to return a substring.

In [ ]:
song_line = "Row, row, row your boat"
what_to_row = song_line[-9:]
print(what_to_row)

your boat


In [ ]:
what_to_do_with_your_boat = song_line[:3]
print(what_to_do_with_your_boat)

Row


In [ ]:
whose_boat_it_is = song_line[14:18]
print(whose_boat_it_is)

your


### <br><br><span style="color:red">Exercise: String indexing and functions

In [ ]:
lyrics = "Lean on me, when you're not strong\nAnd I'll be your friend\nI'll help you carry on"

In [ ]:
print(lyrics)

Lean on me, when you're not strong
And I'll be your friend
I'll help you carry on


<br>Run the two lines of code above to see the string we'll be working with.

<br>Use string indexing to write code to print each line of the song on its own:

In [ ]:
first_line = lyrics[:34]
print(first_line)

Lean on me, when you're not strong


In [ ]:
second_line = lyrics[35:58]
print(second_line)

And I'll be your friend


In [ ]:
third_line = lyrics[58:]
print(third_line)


I'll help you carry on


<br>Write code to replace the new line characters (`\n`) in `lyrics` with a period and a space ". ":

In [ ]:
lyrics_on_one_line = lyrics.replace("\n", ". ")
print(lyrics_on_one_line)

Lean on me, when you're not strong. And I'll be your friend. I'll help you carry on


<br>Write code to capitalize the first letter of every word in `lyrics`:

In [ ]:
lyrics.title()

"Lean On Me, When You'Re Not Strong\nAnd I'Ll Be Your Friend\nI'Ll Help You Carry On"

<br><br>*Does this result look strange to you? We will learn how to fix it later in the workshop.*

### <br><br><span style="color:red">Exercise: More string methods
Python has many other `built in` string method functions. Go to the documentation page for the string object: https://docs.python.org/3/library/stdtypes.html#string-methods. Choose a method function to test out. Does it work the way you thought it would? Are there any arguments that you can pass to the function?